### ISSUE TITLE
List:setChoices rendering issues

### ISSUE THREAD

**Issue by robob27** [DFHack/dfhack#2912]
The test script used to demonstrate these issues can be found [here](https://gist.github.com/robob27/ece52d0adc9310b0018f8c05a8ec7b74). I initially found the issues in a much more complex script so I tried to distill it down to something very simple that could also demonstrate the issues.

If you call `setSelected` in the same frame as the list is rendered, `page_top` is not calculated correctly. This results in the scrollbar and available options being rendered incorrectly:
![image](https://user-images.githubusercontent.com/2104555/219223222-8bad930c-a9f3-40b7-ae71-2388cb8d4f7f.png)

This only happens if there is not enough space to display all of the options, unlike in `gui/confirm` where we have exactly as many confirms as we have available space in the list, so this is not yet a problem. Here is same list choices, with setSelected also called, but with enough space to display all options:

![image](https://user-images.githubusercontent.com/2104555/219223448-6aa062bb-1d09-436d-8e58-cc92003e0828.png)

I believe the issue is only when these both happen in the same frame because if I render the list first, and then do some action to populate the list after, even if there is not enough room to display all options, it renders correctly:

Rendering the list first without any choices:

![image](https://user-images.githubusercontent.com/2104555/219223621-6e91f30f-02ea-4912-95c9-b3de3fa90f59.png)

Then using "Switch lists" to populate the list and set the selected index:

![image](https://user-images.githubusercontent.com/2104555/219223667-f89b7c6b-b136-4001-b8ce-a6f16fb95a55.png)

However, if I now switch between the lists, when I switch to the list that _would_ have enough space to render all of its choices, it does not (One, Two and Three should be visible here):

![image](https://user-images.githubusercontent.com/2104555/219223858-0451c487-10e3-4a3d-a6d3-90e0aea3014e.png)

**Comment by robob27** [comment]
I tried to work on a fix for this, but I didn't get very far by myself. Then I had an idea to ask ChatGPT about it. I fed it the full code for the List widget and described the problem. After some back and forth we ended up with this:

https://github.com/DFHack/dfhack/compare/develop...robob27:list-setselected-fix?expand=1

This _does_ appear to fix the issues described here, but... I didn't come up with it myself entirely, and there was a reason why I asked ChatGPT about it 😂. I figured I would post it and someone who is more comfortable with everything else going on the widget could maybe make use of it to make the ideal fix, assuming this _probably_ isn't exactly it.
